x-pack/plugins/security_solution/public/common/containers/source/index.tsx

```diff
@@ -55,15 +54,34 @@ export const getIndexFields = memoizeOne(
   (newArgs, lastArgs) => newArgs[0] === lastArgs[0] && newArgs[1].length === lastArgs[1].length
 );
 
+/**
+ * HOT Code path where the fields can be 16087 in length or larger. This is
+ * VERY mutatious on purpose to improve the performance of the transform.
+ */
 export const getBrowserFields = memoizeOne(
-  (_title: string, fields: IndexField[]): BrowserFields =>
-    fields && fields.length > 0
-      ? fields.reduce<BrowserFields>(
-          (accumulator: BrowserFields, field: IndexField) =>
-            set([field.category, 'fields', field.name], field, accumulator),
-          {}
-        )
-      : {},
+  (_title: string, fields: IndexField[]): BrowserFields => {
+    // Adds two dangerous casts to allow for mutations within this function
+    type DangerCastForMutation = Record<string, {}>;
+    type DangerCastForBrowserFieldsMutation = Record<
+      string,
+      Omit<BrowserField, 'fields'> & { fields: Record<string, BrowserField> }
+    >;
+
+    // We mutate this instead of using lodash/set to keep this as fast as possible
+    return fields.reduce<DangerCastForBrowserFieldsMutation>((accumulator, field) => {
+      if (accumulator[field.category] == null) {
+        (accumulator as DangerCastForMutation)[field.category] = {};
+      }
+      if (accumulator[field.category].fields == null) {
+        accumulator[field.category].fields = {};
+      }
+      if (accumulator[field.category].fields[field.name] == null) {
+        (accumulator[field.category].fields as DangerCastForMutation)[field.name] = {};
+      }
+      accumulator[field.category].fields[field.name] = (field as unknown) as BrowserField;
```

Comment:
Since, you are building the object through condition, I will make sure to check if you can assign it.
```suggestion
      if (accumulator[field.category].fields[field.name] != null) {
        accumulator[field.category].fields[field.name] = (field as unknown) as BrowserField;
      }
```

Reply:
That condition should never be true because the line above it (L78-L79), if it's null then I create it like so:

```ts
if (accumulator[field.category].fields[field.name] == null) {
  (accumulator[field.category].fields as DangerCastForMutation)[field.name] = {};
}
```

Reply:
Yes, you are right. One more question do we really need to do L78-L79. Why do we need to put an empty object and then assign it? Can we just assign it at the first place?

Reply:
Yes, you are right. One more question, do we really need L78/79? can we just assign it?

Reply:
Yeah I think we can directly assign it, I will change that. Thanks for the 🦅  👁️ 's on this.